### PR TITLE
Save all workspaces after Dynamo crashes

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

Per user report issue and internal report https://github.com/DynamoDS/Dynamo/issues/8913. Fixing the save workflow after Dynamo crashes. Before this PR, the **Save** and **SaveAs** function both default to the currentworkspace which is not always accurate, but is true in most other cases..

After:
User will have a chance to save each workspace with pending changes
SaveWorkspacesAfterDynamoCrash

Two files should be saved as a result of this crash:
image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo 

### FYIs

@Dewb

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@QilongTang how did you make this crash 😉 ?

**Comment by aparajit-pratap** [comment]
> @QilongTang how did you make this crash 😉 ?

I think @QilongTang typed `...` in a code block to make it crash. I think it is filed as an existing issue.

**Comment by aparajit-pratap** [comment]
> I think @QilongTang typed `...` in a code block to make it crash. I think it is filed as an existing issue.
Actually I'm unable to reproduce the crash.

**Comment by QilongTang** [comment]
@aparajit-pratap @mjkkirschner There are steps in https://github.com/DynamoDS/Dynamo/issues/8913

**Comment by QilongTang** [comment]
Running self CI and ready for review

**Comment by QilongTang** [comment]
All unit tests passing

**Comment by mjkkirschner** [comment]
@QilongTang this looks good to me

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 1]
```diff
@@ -1480,7 +1478,34 @@ public void SaveAs(string path, bool isBackup = false)
...
+        /// Save the current workspace to a specific file path. If the file path is null or empty, an
+        /// exception is written to the console.
+        /// </summary>
+        /// <param name="id">Indicate the id of target workspace view model instead of defaulting to 
+        /// current workspace view model. This is critical in crash cases.</param>
+        /// <param name="path">The path to the file.</param>
+        /// <param name="isBackup">Indicates if an automated backup save has called this function.</param>
+        /// <exception cref="IOException"></exception>
+        /// <exception cref="UnauthorizedAccessException"></exception>
+        public void SaveAs(Guid id, string path, bool isBackup = false)
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: How about making this new method private. We can always make it part of the public API if needed later.

**Comment by aparajit-pratap** [comment]
> @aparajit-pratap @mjkkirschner There are steps in #8913

Fixed crash here: https://github.com/DynamoDS/Dynamo/pull/10127

**Comment by aparajit-pratap** [comment]
LGTM with the exception of one comment about making the new method internal.

**Comment by QilongTang** [comment]
@aparajit-pratap Thank you, I made the API internal for now. Will merge once all checks pass

**Comment by mjkkirschner** [comment]
tracked with https://jira.autodesk.com/browse/DYN-1027

**Pull request merged**